File: src/notations/number/knuth/Knuth.cpp
#include "googology/notations/number/knuth/Knuth.hpp"
#include <algorithm>
#include <cctype>
#include <iostream>
#include <memory>

namespace googology {
namespace number {

Capabilities Knuth::capabilities() const {
    Capabilities c;
    c.set(Op::FromString);
    c.set(Op::ToString);
    c.set(Op::Expand);
    c.set(Op::ExpandTo);
    return c;
}

// Deep clone (unique_ptr makes KNode non-copyable).
KNode Knuth::cloneNode(const KNode& n) {
    KNode c;
    c.isVal = n.isVal;
    c.val = n.val;
    c.baseVal = n.baseVal;
    c.height = n.height;
    if (n.exp) c.exp = std::make_unique<KNode>(cloneNode(*n.exp));
    return c;
}

// LaTeX (no numeric evaluation) of a Knuth AST node.
std::string Knuth::tex(const KNode& n) {
    if (n.isVal) return std::to_string(n.val);
    std::string bt = std::to_string(n.baseVal);
    std::string et = n.exp ? tex(*n.exp) : "";
    std::string arr = (n.height == 1)
        ? "\\uparrow"
        : "\\uparrow^{" + std::to_string(n.height) + "}";
    // parenthesize the exponent only when it is itself an arrow term
    bool expHasArrow = n.exp && !n.exp->isVal;
    std::string expPart = expHasArrow ? "(" + et + ")" : et;
    return bt + " " + arr + " " + expPart;
}

// One single rewrite step on the AST. No numeric evaluation: the result is
// always another Knuth AST node (a rewritten form of the same notation).
//   - a ^c 1            = a                (collapse to base)
//   - a ^1 b (= a^b)    : terminal, unchanged
//   - a ^c b (c>1,b>1)  = a ^(c-1) (a ^c (b-1))
// When the exponent is itself an arrow, the step recurses into that exponent.
KNode Knuth::step(const KNode& node) {
    if (node.isVal) return cloneNode(node);            // constant: terminal
    if (node.exp && node.exp->isVal) {
        BigInt B = node.exp->val;
        if (B == 1) return KNode::value(node.baseVal);  // a ^c 1 = a
        if (node.height == 1) return cloneNode(node);   // a^b: terminal
        KNode inner = KNode::arrow(node.baseVal, node.height, KNode::value(B - 1));
        KNode outer = KNode::arrow(node.baseVal, node.height - 1, std::move(inner));
        return outer;
    }
    // recurse into the rightmost sub-term
    KNode r;
    r.isVal = false;
    r.baseVal = node.baseVal;
    r.height = node.height;
    r.exp = std::make_unique<KNode>(step(*node.exp));
    return r;
}

// Recursive-descent parser for right-associative arrow towers.
//   expr := INT ( '^'+ expr )?
// `3^^3^^3` -> arrow(3, 2, arrow(3, 2, value(3)))  ==  3 ↑↑ (3 ↑↑ 3)
// A bare integer (no '^') is a leaf. Any malformed tail throws.
KNode Knuth::parseExpr_(const std::string& s) {
    size_t pos = s.find('^');
    if (pos == std::string::npos) {
        // bare integer / leaf
        if (s.empty()) throw std::invalid_argument("Knuth::string_to_it: empty operand");
        BigInt v = std::stoll(s);
        return KNode::value(v);
    }
    BigInt a = std::stoll(s.substr(0, pos));
    size_t j = pos;
    int c = 0;
    while (j < s.size() && s[j] == '^') { ++c; ++j; }
    if (c == 0) c = 1;                 // defencive: never 0 after finding '^'
    std::string rest = s.substr(j);
    if (rest.empty())
        throw std::invalid_argument("Knuth::string_to_it: dangling arrows (no exponent)");
    KNode exp = parseExpr_(rest);        // exponent is the whole RHS, parsed recursively
    return KNode::arrow(a, c, std::move(exp));
}

void Knuth::string_to_it(const std::string& s) {
    std::string t = s;
    // accept the math symbol ↑ as well (UTF-8 string replace, no narrow-char literal)
    const std::string UP = "↑";
    for (size_t p = t.find(UP); p != std::string::npos; p = t.find(UP, p + 1))
        t.replace(p, UP.size(), "^");
    t.erase(std::remove_if(t.begin(), t.end(), ::isspace), t.end());
    if (t.empty()) throw std::invalid_argument("Knuth::string_to_it: empty input");

    // Recursive-descent: supports right-associative arrow towers
    // (e.g. `3^^3^^3` == `3 ↑↑ (3 ↑↑ 3)`). The old single-shot scan
    // silently dropped everything after the first exponent.
    root_ = parseExpr_(t);
}

std::string Knuth::to_string() const { return tex(root_); }

// Rewrite n single steps, mutating this; return *this (a Knuth object).
Knuth& Knuth::expand(BigInt n) {
    for (BigInt i = 0; i < n; ++i) root_ = step(root_);
    return *this;
}

// Expand until the LaTeX length reaches `len` or the form is stable.
Knuth& Knuth::expand_to(BigInt len) {
    std::string prev = to_string();
    while (true) {
        expand(1);
        std::string cur = to_string();
        if (cur.size() >= static_cast<size_t>(len) || cur == prev) return *this;
        prev = cur;
    }
}

std::istream& operator>>(std::istream& is, Knuth& k) {
    std::string s;
    std::getline(is, s);
    if (!s.empty()) k.string_to_it(s);
    return is;
}

} // namespace number
} // namespace googology

File: googology/notations/number/knuth.py
"""
Knuth up-arrow notation (高德纳箭头).

  a ^c b = a ^ b                 (c = 1)
  a ^c b = a                     (b = 1)
  a ^c b = a ^(c-1) (a ^c (b-1)) (c > 1, b > 1)
"""

from googology.core import Op, Capabilities, Family, Notation, UnsupportedOperation


class KNode:
    """One node of a Knuth up-arrow expression AST."""

    def __init__(self, is_val: bool, val: int = 0, base: int = 0, height: int = 0, exp: 'KNode' = None):
        self.is_val = is_val
        self.val = val
        self.base = base
        self.height = height
        self.exp = exp

    @staticmethod
    def value(v: int) -> 'KNode':
        return KNode(True, val=v)

    @staticmethod
    def arrow(base: int, h: int, e: 'KNode') -> 'KNode':
        return KNode(False, base=base, height=h, exp=e)

    def tex(self) -> str:
        if self.is_val:
            return str(self.val)
        bt = str(self.base)
        et = self.exp.tex() if self.exp else ""
        arr = "\\uparrow" if self.height == 1 else f"\\uparrow^{{{self.height}}}"
        exp_has_arrow = self.exp and not self.exp.is_val
        if exp_has_arrow:
            return f"{bt} {arr} ({et})"
        return f"{bt} {arr} {et}"

    def step(self) -> 'KNode':
        if self.is_val:
            return KNode.value(self.val)
        if self.exp and self.exp.is_val:
            b = self.exp.val
            if b == 1:
                return KNode.value(self.base)  # a ^c 1 = a
            if self.height == 1:
                return KNode.arrow(self.base, self.height, KNode.value(b))
            inner = KNode.arrow(self.base, self.height, KNode.value(b - 1))
            return KNode.arrow(self.base, self.height - 1, inner)
        # recurse into the rightmost sub-term
        return KNode.arrow(self.base, self.height, self.exp.step())

    @staticmethod
    def parse(s: str) -> 'KNode':
        s = s.strip()
        pos = s.find('^')
        if pos == -1:
            if not s:
                raise ValueError("Knuth: empty operand")
            return KNode.value(int(s))
        a = int(s[:pos])
        j = pos
        c = 0
        while j < len(s) and s[j] == '^':
            c += 1
            j += 1
        rest = s[j:].strip()
        if not rest:
            raise ValueError("Knuth: dangling arrows (no exponent)")
        exp = KNode.parse(rest)
        return KNode.arrow(a, c, exp)


class Knuth(Notation):
    """Knuth up-arrow notation."""

    def __init__(self, root: KNode = None):
        self._root = root if root else KNode.value(0)

    # --- taxonomy ---
    @property
    def name(self) -> str:
        return "knuth"

    @property
    def family(self) -> Family:
        return Family.Number

    @property
    def creator(self) -> str:
        return "Donald Knuth"

    @property
    def version(self) -> str:
        return "1"

    # --- capability ---
    @property
    def capabilities(self) -> Capabilities:
        c = Capabilities()
        c.set(Op.FromString)
        c.set(Op.ToString)
        c.set(Op.Expand)
        c.set(Op.ExpandTo)
        return c

    @classmethod
    def from_string(cls, s: str) -> 'Knuth':
        k = cls()
        k.string_to_it(s)
        return k

    # --- string conversion ---
    def string_to_it(self, s: str) -> None:
        t = s.replace("↑", "^").replace(" ", "")
        if not t:
            raise ValueError("Knuth: empty input")
        self._root = KNode.parse(t)

    def to_latex(self) -> str:
        return self._root.tex() if self._root else ""

    # --- operations ---
    def expand(self, n: int = 1) -> None:
        for _ in range(n):
            self._root = self._root.step()

    def expand_to(self, length: int) -> None:
        prev = self.to_latex()
        while True:
            self.expand(1)
            cur = self.to_latex()
            if len(cur) >= length or cur == prev:
                return
            prev = cur